src/sage/groups/perm_gps/permgroup.py

```diff
@@ -1581,6 +1581,93 @@ def smallest_moved_point(self):
         p = self._libgap_().SmallestMovedPoint()
         return self._domain_from_gap[Integer(p)]
 
+    @cached_method
+    def disjoint_direct_product_decomposition(self):
+        r"""
+        Returns the finest partition of the underlying set such that ``self``
+        is isomorphic to the direct product of the projections of ``self``
+        onto each part of the partition. Each part is a union of orbits
+        of ``self``.
+
+        The algorithm is from [CJ2022]_, which runs in time polynomial in
+        `n \cdot |X|`, where `n` is the degree of the group and `|X|` is
+        the size of a generating set, see Theorem 4.5.
+
+        EXAMPLES::
+
+        The example from the original paper::
+
+            sage: H = PermutationGroup([[(1,2,3),(7,9,8),(10,12,11)],[(4,5,6),(7,8,9),(10,11,12)],[(5,6),(8,9),(11,12)],[(7,8,9),(10,11,12)]])
+            sage: S = H.disjoint_direct_product_decomposition();S
+            {{1, 2, 3}, {4, 5, 6, 7, 8, 9, 10, 11, 12}}
+            sage: A = libgap.Stabilizer(H, list(S[0]), libgap.OnTuples);A
+            Group([ (7,8,9)(10,11,12), (5,6)(8,9)(11,12), (4,5,6)(7,8,9)(10,11,12) ])
+            sage: B = libgap.Stabilizer(H, list(S[1]), libgap.OnTuples);B
+            Group([ (1,2,3) ])
+            sage: T = PermutationGroup(gap_group=libgap.DirectProduct(A,B))
+            sage: T.is_isomorphic(H)
+            True
+            sage: PermutationGroup(PermutationGroup(gap_group=A).gens(),domain=list(S[1])).disjoint_direct_product_decomposition()
+            {{4, 5, 6, 7, 8, 9, 10, 11, 12}}
+            sage: PermutationGroup(PermutationGroup(gap_group=B).gens(),domain=list(S[0])).disjoint_direct_product_decomposition()
+            {{1, 2, 3}}
+
+        Counting the number of connected subgroups::
+
+            sage: # optional -- internet
```

Comment:
Oh, sorry, I only saw this now.  I think it's better to mark only the `oeis` line as `optional -- internet`, because the numbers can be computed, and are instructive.